hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java

```diff
@@ -314,31 +328,59 @@ private void initializeWALEntryFilter(UUID peerClusterId) {
     this.walEntryFilter = new ChainWALEntryFilter(filters);
   }
 
-  private void tryStartNewShipper(String walGroupId, PriorityBlockingQueue<Path> queue) {
-    workerThreads.compute(walGroupId, (key, value) -> {
-      if (value != null) {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug(
-              "{} Someone has beat us to start a worker thread for wal group {}",
-              logPeerId(), key);
-        }
-        return value;
-      } else {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug("{} Starting up worker for wal group {}", logPeerId(), key);
+  /**
+   * Synchronized method so that only one item is inserted at a time. Should be the only place that
+   * insertions are performed on the workerThreads data structure.
+   *
+   * @param walGroupId The WAL group ID
+   * @param queue The queue of paths to process
+   */
+  private synchronized void tryStartNewShipper(final String walGroupId,
+      final PriorityBlockingQueue<Path> queue) {
+
+    if (workerThreads.containsKey(walGroupId)) {
+      return;
+    }
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("{} Register worker for wallGroupID {}", logPeerId(), walGroupId);
+    }
+
+    ReplicationSourceShipper worker = createNewShipper(walGroupId, queue);
+
+    ReplicationSourceWALReader walReader =
+        createNewWALReader(walGroupId, queue, worker.getStartPosition());
+
+    // Worker will stop walReader when it stops
+    worker.setWALReader(walReader);
+
+    workerThreads.put(walGroupId, worker);
+
+    // Kick of the WAL reader first to get a head start
+    executorService.submit(walReader);
```

Comment:
Both worker and reader go to same thread pool?

Reply:
Ya.  If I were doing this entire thing from scratch, I would have taken a different approach altogether, but at least there is a one-stop shop to see how many threads are running in total for this effort.  This thread pool is unbounded so no real change in functionality.